### DISCUSSION TITLE
Feedback on the 1.0.0 draft

### DISCUSSION THREAD

**Discussion by earth2marsh** [OAI/Arazzo-Specification#19] [General]
Great to see the [first working draft](https://github.com/OAI/sig-workflows/blob/draft-version/versions/1.0.0.md) landed! I wasn't sure where to put feedback, so starting this discussion, but please adjust me if it should go somewhere else.

1. While this draft is a solid start, the amount of abstraction makes it hard for me to understand what it will look like in practice. I'd benefit from a bunch of examples that accompany the formal rules. May I suggest taking some of the highest priority workflows and imagine them in YAML? That would help me better understand what the authoring experience might be like in practice.
2. The workflows spec would benefit from having a clear purpose at the top of the document, even if that's a placeholder.
3. The word `version` has bitten OpenAPI document authors time and again... consider a multi-word property name here, even it it means diverging from OpenAPI somewhat.
4. Is a workflows document meant to refer to an OpenAPI document? A specific version of that document? Multiple documents?
5. Might each workflow be a separate document?

**Comment by ankrim** [comment]
I agree that we need to add examples, that's one of the efforts that we'll start in parallel to writing the spec.
For now, you can see an example in this discussion https://github.com/OAI/sig-workflows/discussions/10.

**Comment by namdeirf** [comment]
Please provide your use case examples in this discussion #23

**Reply by swaldron58** [reply]
can you suggest a format for use cases? in sig-travel we have looked at several formats. We looking at real world API sequences which takes a while to document in a non specific product or provider way. Like a common hotel booking workflow versus Hilton specifically. what I mean by that is we have some members that can provide the existing sequence of their current API offering. Then there are other high level (not at the API level) documents on how, for example, to exchange digital IDs. 
What will take longer is for the sig-travel team to sort out in, for example and end to end booking flow, how we break this down into procedures. For example is an initial procedure just "shop" followed by others for "book", "pay" and "fulfillment".

**Comment by edwinlzs** [comment]
Hey! Appreciate the work on the spec. Not sure where else to ask clarifications so I'll drop it here.

https://spec.openapis.org/arazzo/latest.html#step-object

It says a step MAY be a call to an API op. I'm reading this as: there might NOT be any calls at all (i.e. its possible that `operationId`, `operationPath` and `workflowId` are all not defined). Would that just be a no-op? What would be the purpose of having such a step?

**Reply by edwinlzs** [reply]
Never mind, I figured it out. I was tripped out by the placement of the parentheses (there might be a mistake in the explanation there), I think what the spec actually meant to say is:

"Describes a single workflow step which MAY be a call to an API operation (OpenAPI Operation Object) or another Workflow Object."

**Comment by orubel** [comment]
Well you know my take. It does not deal with redirection except in the case of external redirects. Multithreaded applications often handle redirection INTERNALLY as well as externally (ie forward/java, loopback/nodejs and stepfunctions/aws). So this is extremely limited, insecure (each api has to talk to another through a round trip outside the DMZ and reissuer the request/response). So ight noiw it is only addressing the slowest, most processor intensive method of doing api call flow.

Single threaded apps cannot do internal redirection due to the placement of their front controller; AWS had to do a work around ('step functions') in order to create this functiuonality since Lambdas is a multithreaded app (Vert.X) while Lambdas are only single threaded instances.

Not ideal in any sense and it may be years before Arazzo/OpenApi add in proper api redirection handling because when I brought this up, the lead was so ill-informed as to accuse me of making this all up. https://github.com/OAI/Arazzo-Specification/issues/261#issuecomment-2397274286

And this is very common usage in alot of products (as well as being addressed by Roy Fieldings dissertation)...
- [ Internal Redirection vs External Redirection ](https://www.codeguage.com/blog/internal-external-redirection)
- [ Internal Redirect in Flask](https://stackoverflow.com/questions/32534051/internal-redirect-in-flask)
- [ Internal redirect in native api tools](https://www.progress.com/documentation/sitefinity-cms/for-developers-create-internal-redirect-page-with-the-native-api)
- [Forward : Java's internal redirect](https://www.baeldung.com/spring-redirect-and-forward)

It's also a part of almost all Api Gateways...
- [A part of EnvoyProxy ](https://github.com/envoyproxy/envoy/issues/19682)
- [and NGINX](https://nginx.org/en/docs/http/ngx_http_internal_redirect_module.html)

The fact that the lead on this thought I was making this up shows that he doesnt fulling understand the api callflow in distributed architectures... only single threaded ones.

**Comment by frankkilcommins** [comment]
v1.0.0 is not longer in `draft` status and has been published. Closing this discussion
